File: MCP/tool-gateway/src/schema-validator.js
import crypto from "node:crypto";
import Ajv from "ajv";

const DEFAULT_AJV_OPTIONS = {
  allErrors: true,
  allowUnionTypes: true,
  coerceTypes: true,
  useDefaults: true,
  removeAdditional: false,
  strict: false,
};

export class SchemaValidator {
  constructor(options = {}) {
    const ajvOptions = options.ajvOptions ?? {};
    this.ajv = new Ajv({ ...DEFAULT_AJV_OPTIONS, ...ajvOptions });
    this.cache = new Map();
  }

  validate(toolId, schema, args = {}) {
    if (!schema || typeof schema !== "object") {
      return args ?? {};
    }

    const validator = this.getValidator(toolId, schema);
    const payload = cloneArgs(args);
    const valid = validator(payload);
    if (!valid) {
      throw new Error(formatAjvErrors(validator.errors));
    }
    return payload;
  }

  getValidator(toolId, schema) {
    const fingerprint = fingerprintSchema(schema);
    const cached = this.cache.get(toolId);
    if (cached && cached.fingerprint === fingerprint) {
      return cached.validate;
    }

    const compiled = this.ajv.compile(cloneSchema(schema, toolId));
    this.cache.set(toolId, { fingerprint, validate: compiled });
    return compiled;
  }
}

function cloneArgs(args) {
  if (!args || typeof args !== "object") {
    return {};
  }
  return deepClone(args);
}

function cloneSchema(schema, toolId) {
  const cloned = deepClone(schema);
  if (cloned && typeof cloned === "object" && !cloned.$id) {
    cloned.$id = `tool://${toolId}`;
  }
  return cloned;
}

function deepClone(value) {
  if (typeof globalThis.structuredClone === "function") {
    return globalThis.structuredClone(value);
  }
  return JSON.parse(JSON.stringify(value));
}

function fingerprintSchema(schema) {
  const serialized = JSON.stringify(schema);
  return crypto.createHash("sha1").update(serialized).digest("hex");
}

function formatAjvErrors(errors) {
  if (!errors?.length) {
    return "Arguments failed schema validation.";
  }
  return errors
    .slice(0, 5)
    .map((error) => {
      const path = error.instancePath || error.schemaPath || "schema";
      return `${path}: ${error.message}`;
    })
    .join("; ");
}

File: MCP/tool-gateway/schema_validator.py
"""JSON-schema validation for tool call arguments."""

from __future__ import annotations

from hashlib import sha1
import json
from typing import Any

from jsonschema import Draft202012Validator


class SchemaValidator:
    def __init__(self) -> None:
        self._cache: dict[str, tuple[str, Draft202012Validator]] = {}

    def validate(self, tool_id: str, schema: dict[str, Any] | None, args: dict[str, Any] | None = None) -> dict[str, Any]:
        payload = _clone_args(args)
        if not isinstance(schema, dict):
            return payload

        validator = self._get_validator(tool_id, schema)
        errors = sorted(validator.iter_errors(payload), key=lambda error: str(error.path))
        if errors:
            detail = "; ".join(_format_error(error) for error in errors[:5])
            raise ValueError(detail or "Arguments failed schema validation.")
        return payload

    def _get_validator(self, tool_id: str, schema: dict[str, Any]) -> Draft202012Validator:
        fingerprint = _schema_fingerprint(schema)
        cached = self._cache.get(tool_id)
        if cached and cached[0] == fingerprint:
            return cached[1]

        cloned = _clone_schema(schema, tool_id)
        validator = Draft202012Validator(cloned)
        self._cache[tool_id] = (fingerprint, validator)
        return validator


def _clone_args(args: dict[str, Any] | None) -> dict[str, Any]:
    if not isinstance(args, dict):
        return {}
    return json.loads(json.dumps(args))


def _clone_schema(schema: dict[str, Any], tool_id: str) -> dict[str, Any]:
    cloned = json.loads(json.dumps(schema))
    if "$id" not in cloned:
        cloned["$id"] = f"tool://{tool_id}"
    return cloned


def _schema_fingerprint(schema: dict[str, Any]) -> str:
    raw = json.dumps(schema, sort_keys=True, ensure_ascii=True)
    return sha1(raw.encode("utf-8")).hexdigest()


def _format_error(error) -> str:
    if error.path:
        path = "/" + "/".join(str(token) for token in error.path)
    else:
        path = error.schema_path[0] if error.schema_path else "schema"
    return f"{path}: {error.message}"
